### PR TITLE
Google Breakpad integration

### PR DESCRIPTION
**Pull request by speqtr**
`!WARNING! This is work in progress. Please, don't merge!`

In order to better handle runtime crashes we need to integrate Google Breakpad.

### CODE REVIEW

**Comment by speqtr** [comment]
**NOTE:** After pushing all the necessary commits, I'll share docs and usage example in comments section of this PR (and later I'll add those to `README.md`).

**Comment by speqtr** [comment]
*UPDATE:* Latest CI build for Linux and macOS look OK. Investigating Android and Windows issues...

**Comment by speqtr** [comment]
*UPDATE:* Breakpad issues for Anroid CI build are fixed.
Windows is still in progress...

**Comment by speqtr** [comment]
*UPDATE:* Windows builds are finally green too 🙂
I'll add a few more commits and some docs/examples tomorrow.

**Comment by speqtr** [comment]
@exploshot @nnian @Xecute0101 @who-biz 

Please, make a simple test:
1. Clone changes from this branch;
2. Modify `src/ConnectivityTool/ConnectivityTool.cpp` main function:
```
// Before
Qwertycoin::Breakpad::ExceptionHandler exceptionHandler;

// After
Qwertycoin::Breakpad::ExceptionHandler exceptionHandler;
Qwertycoin::Breakpad::ExceptionHandler::dummyCrash();
return 0;
```
3. Compile and run `connectivity_tool`, it shoud crash with a similar message:
```
Minidump path: Going to crash...
ERROR: Process crashed! Minidump created.
Minidump dir: /tmp, id: 6393218B-DE63-4478-A73B-53804A2B11A4
```
4. Then in `/tmp` dir (on Linux and macOS) or `C:\Windows\Temp` (on Windows) you should find a file with a name like `6393218B-DE63-4478-A73B-53804A2B11A4.dmp`.

**NOTE:** This is a simple test that will confirm that breakpad exception handler works as expected. 

**NOTE2:** In further updates I'll post information on how to read those generated `*.dmp` files.

**Comment by nnian** [comment]
> 
> 
> @exploshot @nnian @Xecute0101 @who-biz
> 
> Please, make a simple test:
>     1. Compile and run `connectivity_tool`, it shoud crash with a similar message:
> 
> 
> ```
> Minidump path: Going to crash...
> ERROR: Process crashed! Minidump created.
> Minidump dir: /tmp, id: 6393218B-DE63-4478-A73B-53804A2B11A4
> ```
> 
>     1. Then in `/tmp` dir (on Linux and macOS) or `C:\Windows\Temp` (on Windows) you should find a file with a name like `6393218B-DE63-4478-A73B-53804A2B11A4.dmp`.
> 

Yep it works!

**Comment by speqtr** [comment]
@nnian great! And thanks for the update!
PS: Tomorrow I'll post a short doc on how to read those `.dmp` files 🙂

File: lib/CryptoNoteCore/Currency.cpp [hunk 1]
```diff
@@ -141,7 +141,7 @@ bool Currency::getBlockReward(
     uint32_t height,
     uint64_t blockTarget) const
 {
-    assert(alreadyGeneratedCoins <= m_moneySupply);
+    //assert(alreadyGeneratedCoins <= m_moneySupply);
```
Comment (who-biz) [inline lib/CryptoNoteCore/Currency.cpp]: is there a reason for this? unsure of how his relates to Breakpad

File: src/SimpleWallet/SimpleWallet.cpp [hunk 2]
```diff
@@ -3291,6 +3292,8 @@ int main(int argc, char *argv[])
     _CrtSetDbgFlag(_CRTDBG_ALLOC_MEM_DF | _CRTDBG_LEAK_CHECK_DF);
 #endif
 
+    Qwertycoin::Breakpad::ExceptionHandler exceptionHandler;
```
Comment (who-biz) [inline src/SimpleWallet/SimpleWallet.cpp]: Also would be wary about linking this to Wallet and PaymentGate... from a password security standpoint for users. Maybe debug builds only?

File: lib/CryptoNoteCore/Currency.cpp (hunk 1 above)
Comment (speqtr) [reply lib/CryptoNoteCore/Currency.cpp]: True, this one got in this PR by an accident after rebasing on top of some older changes.

File: src/SimpleWallet/SimpleWallet.cpp (hunk 2 above)
Comment (speqtr) [reply src/SimpleWallet/SimpleWallet.cpp]: I think we should keep this `ON` for both Debug and Release builds.

But for security reasons we should never upload those dumps from user machines to anywhere on the internet.

Supporting breakpad in release builds will helps us get more information from users when they post bug reports on GitHub. Basically, we will guide user through readable dump generation process and ask them to post only a stack trace in order to figure out on what like application did actually crash.

Comment (who-biz) [reply src/SimpleWallet/SimpleWallet.cpp]: > I think we should keep this `ON` for both Debug and Release builds.
> 
> But for security reasons we should **never automatically upload** those dumps from user machines to anywhere on the internet.

Didn't realize that automatic uploads weren't enabled by default.  All good on that concern, then.  Thanks for clarifying

**Comment by nnian** [comment]
Is it done?

**Comment by speqtr** [comment]
@nnian oops, sorry, I forgot to revert one line that is unrelated to this PR. I'll do it in a moment. And then, once you see green CI build, you can merge it 🙂

File: lib/CryptoNoteCore/Currency.cpp (hunk 1 above)
Comment (speqtr) [reply lib/CryptoNoteCore/Currency.cpp]: Reverted this unrelated changes in a new commit.

**Comment by speqtr** [comment]
@nnian Travis CI failure can be ignored. It failed due to Travis CI build time limitations. I think this "error" can be ignored.

**Comment by speqtr** [comment]
@nnian @exploshot @who-biz @Xecute0101 I think that we can merge this PR unless you have any extra comments 🙂

**Pull request merged**